src/epcpm/sunspectocsv.py

```diff
@@ -0,0 +1,532 @@
+    sort_models = attr.ib(default=False)
+
+    def gen(self):
+        csv_data = []
+
+        if self.sort_models:
+            children = sorted(
+                self.wrapped.children,
+                key=lambda child: (
+                    0 if isinstance(child, epcpm.sunspecmodel.Model) else 1,
+                    getattr(child, "id", math.inf),
+                ),
+            )
+        else:
+            children = list(self.wrapped.children)
+
+        model_offset = 2  # account for starting 'SunS'
+        for model in children:
+            if isinstance(model, epcpm.sunspecmodel.Table):
+                # TODO: for now, implement it soon...
+                continue
+
+            model_offset += builders.wrap(
+                wrapped=model,
+                csv_data=csv_data,
+                column_filter=self.column_filter,
+                padding_type=self.parameter_model.list_selection_roots[
+                    "sunspec types"
+                ].child_by_name("pad"),
+                parameter_uuid_finder=self.parameter_uuid_finder,
+                model_offset=model_offset,
+            ).gen()
+
+        return csv_data
+
+
+@builders(epcpm.sunspecmodel.Model)
+@attr.s
+class Model:
+    wrapped = attr.ib()
+    csv_data = attr.ib()
+    column_filter = attr.ib()
+    padding_type = attr.ib()
+    model_offset = attr.ib()  # starting Modbus address for the model
+    parameter_uuid_finder = attr.ib(default=None)
+
+    def gen(self):
+        self.wrapped.children[0].check_offsets_and_length()
+
+        overall_length = sum(
+            child.check_offsets_and_length() for child in self.wrapped.children[1:]
+        )
+
+        add_padding = (overall_length % 2) == 1
+        if add_padding:
+            overall_length += 1
+
+        accumulated_length = 0
+
+        rows = []
+
+        model_types = ["Header", "Fixed Block", "Repeating Block"]
+
+        zipped = zip(enumerate(self.wrapped.children), model_types)
+        for (i, child), model_type in zipped:
+            builder = builders.wrap(
+                wrapped=child,
+                add_padding=add_padding and i == 1,
+                padding_type=self.padding_type,
+                model_type=model_type,
+                model_id=self.wrapped.id,
+                model_offset=self.model_offset,
+                parameter_uuid_finder=self.parameter_uuid_finder,
+                address_offset=accumulated_length,
+            )
+
+            built_rows, block_length = builder.gen()
+            accumulated_length += block_length
+            if len(built_rows) > 0:
+                rows.extend(built_rows)
+
+        for i, row in enumerate(rows):
+            if i == 0:
+                row.value = self.wrapped.id
+            elif i == 1:
+                row.value = overall_length
+
+            self.csv_data.append(row.as_filtered_tuple(self.column_filter))
+
+        return overall_length + 2  # add header length
```

Comment:
Same magic number as Alex noted above.

Reply:
I changed the code so that the magic number is no longer used. Gets the +2 from the HeaderBlock. Change in 004cc5dfe8e5765d8f063afd11dec72e1e9f2840